datalayer/src/main/java/com/google/android/horologist/data/apphelper/DataLayerAppHelper.kt

```
@@ -141,18 +145,36 @@ abstract class DataLayerAppHelper(
         config: ActivityConfig
     ): AppHelperResultCode {
         val request = launchRequest { activity = config }
-        val response =
-            registry.messageClient.sendRequest(node, LAUNCH_APP, request.toByteArray()).await()
-        return AppHelperResult.parseFrom(response).code
+        return sendRequestWithTimeout(node, LAUNCH_APP, request.toByteArray())
     }
 
     /**
      * Launch own app on the specified node.
      */
     public suspend fun startRemoteOwnApp(node: String): AppHelperResultCode {
         val request = launchRequest { ownApp = ownAppConfig { } }
-        val response =
-            registry.messageClient.sendRequest(node, LAUNCH_APP, request.toByteArray()).await()
+        return sendRequestWithTimeout(node, LAUNCH_APP, request.toByteArray())
+    }
+
+    /**
+     * Attempts a [MessageClient#sendRequest] with a timeout, covering both the GMS Timeout that may
+     * occur (after 1 minute?) but also a timeout specified by Horologist, as 1 minute is often too
+     * long.
+     */
+    protected suspend fun sendRequestWithTimeout(node: String, path: String, data: ByteArray): AppHelperResultCode {
+        val response = try {
+            withTimeout(MESSAGE_REQUEST_TIMEOUT_MS) {
+                registry.messageClient.sendRequest(node, path, data).await()
+            }
+        } catch (timeoutException: TimeoutCancellationException) {
+            return AppHelperResultCode.APP_HELPER_RESULT_TIMEOUT
+        } catch (e: ApiException) {
+            if (e.statusCode == CommonStatusCodes.TIMEOUT) {
+                return AppHelperResultCode.APP_HELPER_RESULT_TIMEOUT
```

This worries me that it will be silently ignored. Should we either throw a well known timeout exception here, or enforce checking result?

---

Or turn other errors into a 7th value?

---

Yeah, I agree with the issue of being silently ignored. Maybe: (1) throw the well-known timeout exception, (2) don't catch ApiException at all, so the developer needs to handle the timeout exception?

---

From your own use, which ones would you prefer to deal with?

---

Reading a bit more about CheckResult I've gone that route for now.